java/core/src/main/java/co/worklytics/psoxy/gateway/impl/CommonRequestHandler.java

```diff
@@ -122,17 +124,28 @@ public HttpEventResponse handle(HttpEventRequest request) {
             return healthCheckResponse.get();
         }
 
-        // re-write host
-        URL targetUrl = buildTarget(request);
-        String relativeURL = URLUtils.relativeURL(targetUrl);
+        // parse requested URL, re-writing host/etc
+        String requestedTargetUrl = parseRequestedTarget(request);
+
+        // if requested target URL has tokenized components, reverse
+        String clearTargetUrl = reverseTokenizedUrlComponents(requestedTargetUrl);
+
+        boolean tokenizedURLReversed = Objects.equals(requestedTargetUrl, clearTargetUrl);
+
+        String reversalOutcome = tokenizedURLReversed ? "No reversible tokens detected in URL." : "Detected tokens reversed.";
+
+        URL targetUrl = new URL(clearTargetUrl);
+
+        // avoid logging clear URL outside of dev
+        URL toLog = envVarsConfigService.isDevelopment() ? new URL(requestedTargetUrl) : targetUrl;
```

Comment:
isn't this flipped? you are logging `targetUrl` (that is the clear one) if not in development...

In any case, I'd log always the same, the requestedTargetUrl
And add an extra log call for development

if (development) {
  log(targetUrl);
}